### ISSUE TITLE
Reproduce prod environment on staging

### ISSUE THREAD

**Issue by Jackseed** [blockframes/blockframes#3162]
## Description
Staging has diverged from prod and is no longer a trusted test environment. 
In order to test properly we need to have on staging an up-to-date firestore & firestorage taken from prod. We also need to anonymize prod firestore beforehand by changing:
- Org name (`orgs.denomination.full` & `orgs.denomination.public`)
- Org contact email address (`orgs.email`)
- User name (`user.firstName` & `user.lastName`)
- User email (`user.email`)

Anonymizing prod protects users' data privacy and avoid staging to spam emails to real users. 

In order to be easily readable for us (and thus better testing) we used these conventions when generating the fake data:
- org contact email address was derived from org name (contact@orgName.fake.cascade8.com)
- user email was derived from user name (userName@orgName.fake.cascade8.com)

These conventions should be updated to use real email address owned by cascade8 to avoid our Sendgrid email score to be polluted by the tests. 
E.g. use something like fakeAccounts+orgName-contact@cascade8.com & fakeAccounts+userName-orgName@cascade8.com

The process should be this:
1. export prod db & storage
2. anonymize prod db
3. import anonymized prod db & storage
4. erase and recreate firebase authentication users based on anonymize prod db

Best would be to be able to activate that script whenever we want and/or to activate it on each staging deploy. 

## Current tools
- Functions exist to backup & restore firestore following this url: projectHostingUrl/admin/data/backup & projectHostingUrl/admin/data/restore. [Here](https://www.notion.so/cascade8/Data-import-backup-c3279edb834f4a25b130adaecacef0d4) is an article to set it up on a project.
When restore function is used, authentication is reset to match firestore
- [Notion article](https://www.notion.so/cascade8/Bucket-storage-backup-restore-000b1dbb9df8453d807377e158a2450b) to use gs util to import & export firestorage; Google cloud also [exposes functions](https://console.cloud.google.com/transfer/cloud?project=blockframes&authuser=1) to do the same
- Anonymizing prod was done with a Python script 

## Acceptance criteria
- [ ] Update backup & restore db functions if needed ([ressource](https://dev.to/mikgross/firebase-firestore-migrations-olg) from François
- [ ] Add a storage backup & restore to the db functions
- [ ] Include a script to anonymize db
- [ ] Add it on the deploy staging flow and/or as a callable script

**Comment by george43g** [comment]
"Staging has diverged from prod and is no longer a trusted test environment"

You've described how the process works well, and I've taken a look at the code. But you haven't described clearly what the current problem is - is it with the current data, or the process? Which parts of the process? Which part of the data? I can take a look at it myself, but I may not catch all of the issues that you encountered. I've taken a look in the CI, there's a recent data backup there from 11th June. Is this process no longer working? Is there a log of the failure on circleCI?

**Comment by Jackseed** [comment]
The current problem is that we have distinct flows to do the 4 described points in the process. We need 1 script that does it all + anonymizing script must be written again as it's Python and we don't have it anymore.
Backup / import firestore & firestorage works fine.

**Comment by george43g** [comment]
First step is to clean up existing backup process.

I am getting rid of zapier and using built in scheduling. I will examine this process and get the backups working cleanly. Then I will progress further on this issue. This will also close #3245

**Comment by george43g** [comment]
The zapier trigger works by hitting a URL - it does not directly call a function, it submits a form on https://app.blockframes.io/admin/data/backup with a hardcoded password. This triggers a function which makes a DB backup and saves it in storage (either its own or maybe staging?).

This is the first part of the process. 

*Improvements*
- At this stage we will no longer use Zapier - functions can be scheduled declaratively from within the functions code itself. This removes one point of failure/complexity.
- The code that does the actual backing-up should be decoupled from function triggers and accessible on its own
- ultimately the majority of code should be in libs - app specific libs can go in a folder with that app name. Apps are meant to be minimal code to pull libraries together and configure build target.
- so we should be easily able to write a function that runs a backup and stores it in storage.
- should it even be a cloud function on a trigger? Could it be a CI job that gets triggered? Should it be possible to trigger locally? So let's think from first principle - we take firestore and storage data, timestamp it, store it. Source is prod. easiest destination not to get lost or confused is probably a storage bucket on the same firebase project. So each project should regularly backup its DB to its own storage bucket, via a scheduled function. This seems to be the cleanest approach. This function should be able to be triggered locally or manually.
- The next question is access. How do we allow devs to interact with it? I think there are two possibilities - a script/command/cli tool, or directly accessing web interface. The most flexible solution would be to allow both.
- The way scheduling works is via an adapter over a pubsub function. This means we can both schedule it and also manually call it using pubsub. This will enable us to call the backup from web interface, from CLI, and automatically as well. This is the most flexible solution. The same backup code could also be run locally. So the backup code itself should be stored in a lib.
- backup code already exists, already stored in its own module. Good.

*Status*
- simple function written to trigger database backup, using existing function.
- want to review code for bugs

Have opened PR to show progress: #3258

*plan*
- refactor functions to do with backing up, restoring, clearing - auth, db, storage, into own libs so they can be used from tests, scripts, and more
- review schedule code and existing functions. I suspect there already exist many better ways to backup DB since these were written (choose from npm lib, docs provided func, existing func, CLI code)
- create multiple interfaces into the code - wether it be CLI, web based, script based - will discuss which is a priority if any.
- there should be easy CRUD and access control for the backups. This means security rules and authentication.
- this should easily fit within the process of generating golden data - so how would this flow in a process where someone requests a certain golden data? The issue here is that all of a sudden, we have cross-project requirements for code.
- should timestamped backups be stored alongside their anonymised counterpart?
- how do we manage moving backups from one environment to another? Like how do we load the anon data into another environment? Be it storage, Firestone, auth?
- should we backup/restore auth or generate it?
- ideally, every part of the system is a snapshot - this is the purpose of staging.
- each snapshot should be deployable with a click.
- we need code that can create the backups, and code that can take a backup and deploy it to an environment.
- maybe each project will have this function - it can be called with firebase project config data that it's deploying to. Maybe also via pubsub, to allow any interface to access/command this function.
- what if devs want to run granular operations? like, just backup a db, or just anonimise it, or just restore a raw backup to demo env 2 and a newer backup to prod, for example?

To be continued...

**Comment by george43g** [comment]
Priority tasks and plan as of today:
- complete PR about scheduling backup function without zapier
- open new PR about translating anonymisation script to TS from python
- after that, open the next PR about automating this, and the following:
```
node run .../prepareForTesting
It should :
download the latest anonymized database from blockframes-ci storage (for example).
run algolia changes
reset my authentication / db
run the migration
```

*More issues:*
```
For me the urgency are (in order)
make sure we can continue creating anonimized golden data by hand (updating the python script to typescript that we can run by hand)
Automate the creation of this anonimized golden data (in the CI for example after each backup)
Create a CLI to get the latest golden data before we start coding and append it to prepareForTesting
Remove unnecessary dependancies (Zapier, ...)
Clean up (move code to the right place, ...)
```
"currently we need to download by hand the latest backup, run the anonimisation script and then push back the content in our storage before running the script"

**Comment by GrandSchtroumpf** [comment]
Let's divide this issue in several PRs : 
- [ ] Translate python script into Typescript so we can use it by hand.
- [ ] Automate the creation of a new anonymized database (based on backup) every week through the CI.
- [ ] Update the `prepareForTesting` script to download the latest anonymized database.
- [ ] Create npm script to build `backend-ops` & run `prepareForTesting`
- [ ] Move files at a better place (use `tools` folder for ops & libs for reusable code).
- [ ] Get rid of unnecessary dependancies (Zapier, ...)
- [ ] Clean up

**Comment by george43g** [comment]
Progress:
- the plan of loading prod db into emulator and making the changes there is put on hold. Too many bugs/errors and issues in getting it to work.
- currently working on integrating it into `prepareForTesting` script as that automatically loads a new DB from storage
- anonimisation will be applied like a "migration"

This is the simplest way to move forward. 
Backup and restore of DB seems to always be done by the admin cloud function. Even local scripts (backend-ops) will call a URL of your own env to call your admin function to do so.
The migrations however, connect directly to DB to perform the changes.

There is a "default" user's file that is also applied in the `prepareForTesting` process. This is because the process of generating user accounts from the firestore backup has not been automated yet. This has also been a problem.

There are many parts to this process that can be improved. However, I want to initially focus on the highest priority items - figure out the "critical path" to getting new anonimised prod databases available to devs.

**Comment by george43g** [comment]
New plan:
Download backup
shutdown cloud functions to be sure it doesn't send mails (do via maintenance mode)
Upload backup to database
Modify each user
add anon user to Authentication
setup cloud function again

**Comment by george43g** [comment]
This is a big issue that has changed over time. I think it's probably worth closing this one and starting fresh considering how much we've learned and how much has changed over time. I will save the most important parts from here and enter them into a fresh issue #3611
